File: BOJ/python3/11659.py
'''
구간 합 구하기 4
https://www.acmicpc.net/problem/11659
'''
import sys
import time
start = time.time()  # 시작 시간
sys.stdin = open("input.txt", "r")  # input.txt 파일로 입력

N, M = map(int, input().split())
nums = list(map(int, input().split()))

R = []
for _ in range(M):
    i, j = map(int, sys.stdin.readline().split())
    R.append((i, j))

# 누적합 계산
accs = [0] * (N+1)
acc = 0
for n in range(N):
    acc += nums[n]
    accs[n+1] = acc

# 구간합 계산
for m in range(M):
    i, j = R[m]
    ans = accs[j] - accs[i-1]
    print(ans)

# 현재시각 - 시작시간 = 실행 시간(sec)
print("--------------------")
print(f"Elapsed Time: {round(time.time() - start, 3)}s")

File: BOJ/nodejs/11659.js
/*
구간 합 구하기 4
https://boj.kr/11659
 */
const start = new Date().getTime();

const input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .split("\n");

const [N, M] = input[0].split(" ").map(Number); // [수의 개수, 합을 구해야 하는 횟수]
const nums = input[1].split(" ").map(Number);
const ranges = input.slice(2).map((ele) => ele.split(" ").map(Number));

const DP = new Array(N + 1).fill(0);
let acc = 0;
for (let i = 0; i < N; i++) {
  acc += nums[i];
  DP[i + 1] = acc;
}

const res = new Array(M);
for (let m = 0; m < M; m++) {
  const [l, r] = ranges[m];
  res[m] = DP[r] - DP[l - 1];
}
const ans = res.join("\n");

console.log(ans);

const end = new Date().getTime();
console.log("--------------------");
console.log(`Elapsed Time: ${(new Date().getTime() - start) / 1000}s`);
